cmake: use the warnings_as_errors flag for cpp files

fabiobaltieri opened the pull request:
C++ file compilation is actually missing the warning-as-error handling, causing warnings in build files to be unnoticed in CI. Add a flag to handle them as well.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/78348

Pushed on https://github.com/zephyrproject-rtos/zephyr-testing/tree/vfabio-testing-branch as well

fabiobaltieri committed:
cmake: use the warnings_as_errors flag for cpp files

C++ file compilation is actually missing the warning-as-error handling,
causing warnings in build files to be unnoticed in CI. Add a flag to
handle them as well.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

nashif commented:
and now to all those build issues we missed previous and exposed through this change....

yperess commented:
@fabiobaltieri would you like me to make these fixes in the tree?

fabiobaltieri commented:
> @fabiobaltieri would you like me to make these fixes in the tree?

That'd be cool if you have time, I'm out for the next couple of days.

nordicjm commented:
Just noticed this is a duplicate of https://github.com/zephyrproject-rtos/zephyr/pull/69226

fabiobaltieri committed:
mspi: add a dummy byte to struct mspi_timing_cfg

Add a dummy byte to struct mspi_timing_cfg, for C++ compatibility, same
as what's done in include/zephyr/arch/structs.h,
include/zephyr/arch/arm/structs.h, include/zephyr/kernel/thread.h and
others.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
tests: i2c_emul: fix sign comparison warning

Fix the compiler error:

comparison of integer expressions of different signedness: ‘const int’
and ‘unsigned int’

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Thanks @nordicjm, borrowed some commits from there and added few more, still not done though but I could use a CI run, moving back to draft.

fabiobaltieri committed:
tests: cpp: add a pm device reference

On some architecture (qemu_x86) this tests generate an unused reference
warning for the PM device structure, as nothing in the structure macro
definition is referencing itself.

Reference the struct in the device define, keep the compiler happy.

Also change one of the device priority to dodge a

'__device_dts_ord_8' causes a section type conflict with
'__device_dts_ord_9'

error with llvm.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
tests: i2c_emul: drop the array designators

Drop the array dasignators as they generate a

warning: array designators are a C99 extension

warning when compiling with clang. Not too sure if this is an actual
problem or if the warning should be disabled instead but since it's just
a test it should be fine to just drop them.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

yperess commented:
Looks like you got them, I pushed https://github.com/zephyrproject-rtos/zephyr/pull/78614 the other day and was able to pass CI today. Yours is better structured so I'm happy to stick with this PR.

fabiobaltieri commented:
Still not there though, one more odd one will look into it tomorrow, not sure how https://github.com/zephyrproject-rtos/zephyr/pull/78614 passes CI, I bumped into more here, must be running a different set of tests, sure more will come up on the push run.

fabiobaltieri committed:
device_mmio: use explicit unsigned literals for addresses

Add an explicit unsigned literal suffix (U) for addresses obtained from
devicetree, this is to work around a specific warning:

zephyrproject/zephyr/include/zephyr/arch/arm/cortex_a_r/timer.h:40:2:
warning: this decimal constant is unsigned only in ISO C90

When compling

west build -p -b qemu_cortex_a9 -T tests/lib/cpp/cxx/cpp.main.cpp98

Argueably all devicetree addresses and sizes should be unsigned but this
is enough to make CI pass, may look into changing the device.h macro as
a followup.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
One more, let's see how it goes, pushed on https://github.com/zephyrproject-rtos/zephyr-testing/tree/vfabio-testing-branch as well just to anticipate any surprise on the push test plan.

cfriedt reviewed (dismissed):
cpp changes look fine

fabiobaltieri commented on include/zephyr/sys/device_mmio.h:
```diff
@@ -63,13 +63,13 @@ struct z_device_mmio_rom {
 #define Z_DEVICE_MMIO_ROM_INITIALIZER(node_id) \
 	{ \
-		.phys_addr = DT_REG_ADDR(node_id), \
+		.phys_addr = UINT32_C(DT_REG_ADDR(node_id)), \
```
tempting to have this as part of `DT_REG_ADDR` (and potentially `DT_REG_SIZE` too) macro itself but that would be a substantial change, definitely worth its own PR, I propose to patch this one up for now. Somewhat surprised CI does not hit more cases of this warning though.

fabiobaltieri replied:
it's a quick fix, that's pretty much what I meant in the comment above

teburd replied:
Please create an issue for this then, if its meant to be revisited

jeppenodgaard replied:
There already exist an issue:
https://github.com/zephyrproject-rtos/zephyr/issues/53505
I gave it a shot, but gave up:
https://github.com/zephyrproject-rtos/zephyr/pull/70761

fabiobaltieri replied:
@jeppenodgaard what was wrong with that? CI errors? I was considering giving it a go as well

jeppenodgaard replied:
@fabiobaltieri see this closing comment:
https://github.com/zephyrproject-rtos/zephyr/pull/70761#issuecomment-2295972704

teburd commented on tests/drivers/i2c/i2c_emul/src/emulated_target.cpp:
```diff
@@ -42,7 +42,7 @@ struct i2c_target_callbacks emulated_callbacks[FORWARD_COUNT] = {
 
 #define DEFINE_EMULATED_TARGET_CONFIG(node_id, prop, n)                                            \
-	[n] = {                                                                                    \
+	{                                                                                          \
```
This looks like its doing something completely different than it was before? Confused by this change, why is "[n] = " no longer needed, why was it needed before?

fabiobaltieri replied:
no idea why it was introduced originally, I guess to do something with that `n` variable? regardless, the compiler does not like it, the alternative would be to shut off that warning, happy with either. @yperess?

teburd commented on tests/drivers/i2c/i2c_emul/src/test_forwarding_pio.cpp:
```diff
@@ -23,7 +23,7 @@ ZTEST(i2c_emul_forwarding, test_write_is_forwarded)
 	// Try writing some values
 	for (uint8_t data = 0; data < 10; ++data) {
-		const int expected_call_count = 1 + data;
+		const unsigned int expected_call_count = 1 + data;
```
I take it the .call_count is unsigned?

fabiobaltieri replied:
`fff.h` agrees

teburd commented on tests/lib/cbprintf_package/src/main.c:
```diff
@@ -923,7 +923,7 @@ ZTEST(cbprintf_package, test_cbprintf_package_convert_static)
 
 	clen = cbprintf_package_convert(spackage, slen, NULL, 0, copy_flags, NULL, 0);
-	zassert_true(clen == slen + sizeof(test_str1) + 1/*null*/ - 2 /* arg+ro idx gone*/);
+	zassert_true(clen == slen + (int)sizeof(test_str1) + 1 /*null*/ - 2 /* arg+ro idx gone*/);
```
This cast looks suspicious to me, its a string length comparison but I guess we need signed casting because... why?

fabiobaltieri replied:
there's no single solution I suppose, Yuval proposed a different one here https://github.com/zephyrproject-rtos/zephyr/pull/78614/files#diff-cd2a686f83bf01b5ca4a68be63cb5c470f0413e7609411bba864eeb8e25fe885R927, I don't particularly care, just pulled this from a PR that was already approved, feel free to suggest any alternative that you think make more sense, I suppose it's just a matter of making the types coherent

fabiobaltieri replied:
cool, picked Yuval version and recredited the commit

teburd commented on tests/lib/cpp/cxx/src/main.cpp:
```diff
@@ -106,6 +106,7 @@ BUILD_ASSERT(ARRAY_SIZE(foos) == 5, "expected 5 elements");
 /* Check that SYS_INIT() compiles. */
 static int test_init(void)
 {
+	(void)foos;
```
why do we need a random cast here? this doesn't look to be doing anything at all?

fabiobaltieri replied:
isn't this how one ignores return values? odd though, it's marked as __maybe_unused, I can try to drop it, again I pulled it from the other PR, can't recall if this was erroring out or not

teburd replied:
Sure but this isn't calling a function is it? its casting a value to (void) which would be a do nothing operation I'd think? This is really confusing to read.

fabiobaltieri replied:
let me try and drop it, may not be needed after all for whatever reason

fabiobaltieri replied:
alright gone, let's see if CI is still happy, can't repro locally so it may be alright

teburd reviewed (changes requested):
Some answers I think are needed for some of the more suspicious changes to appease the compiler, or perhaps some inline comments for the really odd ones (void)foo; looks like a no-op

teburd commented on tests/lib/cpp/cxx/src/main.cpp:
```diff
@@ -153,3 +153,12 @@ DEVICE_DT_DEFINE(DT_NODELABEL(test_dev1_dfr), NULL, NULL, NULL, NULL, POST_KERNE
+static int test_pm_device(void)
+{
+	/* Reference the data structure to avoid an "unused-variable" warning */
+	(void)PM_DEVICE_DT_GET(DT_NODELABEL(test_dev0_boot));
```
Don't we have something like a macro/attribute to inform the compiler to ignore the unused checks? This function that does nothing but cast is... something.

fabiobaltieri replied:
of course but we can't really apply it in this case due to how `PM_DEVICE_DT_DEFINE` expands

teburd replied:
I don't get how this actually does anything useful, to me it looks like the compiler is complaining and that now results in us creating this bodge to make it happy. That really feels wrong on some guttural level to me.

fabiobaltieri replied:
Sure, what alternative are you proposing?

teburd replied:
I guess avoiding C++ warnings that conflict with C warnings somehow? I don't know C++ that well anymore but there's like a 100 warning flags for the compiler and tons of attributes and such, none of them help us avoid this sort of thing?

fabiobaltieri replied:
maybe I can instantiate a device instead of a dummy function with a void reference, hang on

fabiobaltieri replied:
done, makes a lot more sense honestly

fabiobaltieri commented on include/zephyr/drivers/mspi.h:
```diff
@@ -204,7 +204,10 @@ enum mspi_timing_param {
-
+#ifdef __cplusplus
+	/* For C++ compatibility. */
+	uint8_t dummy;
```
This is how it's done in multiple other locations, I mentioned that in the commit message, not trying to pretend I know why but I don't really feel like getting creative on these quirks.

fabiobaltieri committed:
tests: lib: cbprintf_package: fix compiler warning

Fix a type mismatch compiler warning in the c++ tests.

Signed-off-by: Yuval Peress <peress@google.com>
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

teburd commented:
Dropped the XS tag as its a bit more than one line, hopefully no one is offended by this.

fabiobaltieri commented:
Ok this is interesting now:

```console
$ export ZEPHYR_TOOLCHAIN_VARIANT=llvm
$ west build -p -b native_sim -T tests/lib/cpp/cxx/cpp.main.minimal
...
zephyr/tests/lib/cpp/cxx/src/main.cpp:154:1: error: '__device_dts_ord_8' causes a section type conflict with '__device_dts_ord_9'
```

Only on LLVM, seems to not like the two devices one with PM data one without at the same priority, this is the expanded definition if anyone wants to have a go at figuring what is going on

```c
static __attribute__((__aligned__(__alignof(struct device_state)))) struct device_state __devstate_dts_ord_9 __attribute__((__section__(".z_devstate"))); static_assert(sizeof("\"dev1\"") <= 48U, "" "DEVICE_NAME_GET(\"dev1\")" " too long"); static const struct device_dt_nodelabels __dev_dt_nodelabels_dts_ord_9 = { .num_nodelabels = 1, .nodelabels = { "test_dev1_dfr", }, }; static const struct device_dt_metadata __dev_dt_meta_dts_ord_9 = { .nl = &__dev_dt_nodelabels_dts_ord_9, };; const __attribute__((__aligned__(__alignof(struct device)))) struct device __device_dts_ord_9 __attribute__((section("." "_device" "." "static" "." "3_33_"))) __attribute__((__used__)) = { .name = "dev1", .config = (__null), .api = (__null), .state = (&__devstate_dts_ord_9), .data = (__null), .pm_base = (__null), .dt_meta = &__dev_dt_meta_dts_ord_9, }; static const __attribute__((__aligned__(__alignof(struct init_entry)))) struct init_entry __attribute__((__used__)) __attribute__((__section__(".z_deferred_init"))) __init___device_dts_ord_9 = { .init_fn = {.dev = (__null)}, .dev = &__device_dts_ord_9, };;

static int fake_pm_action(const struct device *dev,
  enum pm_device_action pm_action) { return -1; }
; static struct pm_device __pm_device_dts_ord_8 = { .base = { .flags = (((0 << PM_DEVICE_FLAG_WS_CAPABLE) | (0 << PM_DEVICE_FLAG_RUNTIME_AUTO) | (0 << PM_DEVICE_FLAG_PD)) | (0 ? (1UL << (PM_DEVICE_FLAG_ISR_SAFE)) : 0)), .state = PM_DEVICE_STATE_ACTIVE, .action_cb = fake_pm_action, .domain = __null, }, .lock = { .wait_q = { { {(&(&(__pm_device_dts_ord_8.lock).wait_q)->waitq)}, {(&(&(__pm_device_dts_ord_8.lock).wait_q)->waitq)} } }, .count = (1), .limit = (1), }, .event = { .wait_q = { { {(&(&__pm_device_dts_ord_8.event.wait_q)->waitq)}, {(&(&__pm_device_dts_ord_8.event.wait_q)->waitq)} } }, .events = 0 }, };

static __attribute__((__aligned__(__alignof(struct device_state)))) struct device_state __devstate_dts_ord_8 __attribute__((__section__(".z_devstate"))); static_assert(sizeof("\"dev0\"") <= 48U, "" "DEVICE_NAME_GET(\"dev0\")" " too long"); static const struct device_dt_nodelabels __dev_dt_nodelabels_dts_ord_8 = { .num_nodelabels = 1, .nodelabels = { "test_dev0_boot", }, }; static const struct device_dt_metadata __dev_dt_meta_dts_ord_8 = { .nl = &__dev_dt_nodelabels_dts_ord_8, };; const __attribute__((__aligned__(__alignof(struct device)))) struct device __device_dts_ord_8 __attribute__((section("." "_device" "." "static" "." "3_33_"))) __attribute__((__used__)) = { .name = "dev0", .config = (__null), .api = (__null), .state = (&__devstate_dts_ord_8), .data = (__null), .pm_base = (((struct pm_device_base *)&__pm_device_dts_ord_8)), .dt_meta = &__dev_dt_meta_dts_ord_8, }; static const __attribute__((__aligned__(__alignof(struct init_entry)))) struct init_entry __attribute__((__used__)) __attribute__((__section__( ".z_init_" "POST_KERNEL" "33""_" "00008""_"))) __init___device_dts_ord_8 = { .init_fn = {.dev = (__null)}, .dev = &__device_dts_ord_8, };;
```

I'll change the priority for now, no doubt it'll come to bite me back down the road anyway.

fabiobaltieri commented:
Dug a bit more, seems like having that field set moves the device structure from rodata to data, with different prio and no pm:

```
        0x00000000 - 0x000000fe is .text
...
        0x00000130 - 0x00000180 is .bss
...
        0x00000388 - 0x000003a0 is .rodata
        0x000003a0 - 0x000003bc is ._device.static.3_33_
        0x000003bc - 0x000003c4 is .z_deferred_init
        0x000003c4 - 0x000003e0 is ._device.static.3_34_
        0x000003e0 - 0x000003e8 is .z_init_POST_KERNEL33_00008_
        0x000003e8 - 0x000003ec is .init_array
```

with pm:

```
        0x00000000 - 0x00000102 is .text
...
        0x00000134 - 0x00000184 is .bss
...
        0x0000038c - 0x000003a4 is .rodata
        0x000003a4 - 0x000003c0 is ._device.static.3_33_
        0x000003c0 - 0x000003c8 is .z_deferred_init
        0x000003c8 - 0x00000428 is .data
        0x00000428 - 0x00000444 is ._device.static.3_34_
        0x00000444 - 0x0000044c is .z_init_POST_KERNEL33_00008_
        0x0000044c - 0x00000450 is .init_array
```

no idea why it's doing that, does not look right to me but I'll just add it to the ever growing list of c++ stuff I don't understand.

teburd commented:
> Dug a bit more, seems like having that field set moves the device structure from rodata to data, with different prio and no pm:
> 
> ```
>         0x00000000 - 0x000000fe is .text
> ...
>         0x00000130 - 0x00000180 is .bss
> ...
>         0x00000388 - 0x000003a0 is .rodata
>         0x000003a0 - 0x000003bc is ._device.static.3_33_
>         0x000003bc - 0x000003c4 is .z_deferred_init
>         0x000003c4 - 0x000003e0 is ._device.static.3_34_
>         0x000003e0 - 0x000003e8 is .z_init_POST_KERNEL33_00008_
>         0x000003e8 - 0x000003ec is .init_array
> ```
> 
> with pm:
> 
> ```
>         0x00000000 - 0x00000102 is .text
> ...
>         0x00000134 - 0x00000184 is .bss
> ...
>         0x0000038c - 0x000003a4 is .rodata
>         0x000003a4 - 0x000003c0 is ._device.static.3_33_
>         0x000003c0 - 0x000003c8 is .z_deferred_init
>         0x000003c8 - 0x00000428 is .data
>         0x00000428 - 0x00000444 is ._device.static.3_34_
>         0x00000444 - 0x0000044c is .z_init_POST_KERNEL33_00008_
>         0x0000044c - 0x00000450 is .init_array
> ```
> 
> no idea why it's doing that, does not look right to me but I'll just add it to the ever growing list of c++ stuff I don't understand.

Unsolved Mysteries of C++, maybe one of the people who wished to add C++ can help here, I have no idea either.

fabiobaltieri commented:
Yeah, anyway CI seems to be happy again so at least the workarounds are working around.

teburd reviewed (approved):
Noted issue exists for the unsigned dt stuff, all other mysterious things look better. Seems good to me.
